docs/design/fabric-v2+/interfaces.ecc-tlcc-channel.md

```diff
@@ -0,0 +1,541 @@
+  `<chaincode_id, enclave_id, tlcc_mrenclave, channel_id, channel_hash>`
+  as returned by the `tl_session_get_*` functions and their view on their
+  own identity.
+- lastly, the module ensures that for the duration of a single
+  sessions, the interaction is between the same set of enclave
+  _instances_. Or put differently, restarting a tlcc or ecc enclave
+  will not allow to hijack a session and confuse the existing
+  enclaves. More concretely, it means that all session state and in
+  particular the session keys reside only ephemerally in private
+  memory but are not persisted.
+
+## Internals
+
+### Crypto
+
+The actual message authentication code is abstracted below as
+mac/MAC.  For actual choice we have following considerations:
+- SGX SDK provides us with sgx_key_128bit_t as only choice
+  (underlying assumption seems that one would use the key for AES128-GCM).
+- that keylength would match CMAC. However, CMAC is not currently offered by PDO crypto.
+- HMAC-SHA256 would from security perspective ask for 256-bit key
+  (but "technically" supports shorter keys)
+
+This would lead to following options:
+1. add CMAC to pdo crypto
+2. use HMAC with 128bit keys (and reduced security)
+3. use HMAC & use a KDF to stretch the 128bit to 256bit (and better security over (2))
+
+
+### Session State
+
+#### ECC
+```c++
+
+    tl_cc2scc_handler_t cc2scc_handler;
+```
+
+#### ECC_Enclave
+```c++
+    #include <sgx_dh.h>
+
+    // interal session context
+    enum tl_ecc_state {INIT1, INIT2, ACTIVE, CLOSING, CLOSED} tl_ecc_state_t;
+
+    struct tl_ecc_ctx {
+        session_id_t session_id,
+        tl_ecc_state_t state,
+        sgx_dh_session_t dh
+        sgx_key_128bit_t session_key
+        char* channel_id,
+        channel_hash_t channel_hash,
+        char* chaincode_id,
+        char* enclave_id,
+        sgx_dh_session_enclave_identity_t tl_identity,
+    } tl_ecc_ctx_t;
+```
+
+
+#### TLCC_Enclave
+
+```c++
+  #include <sgx_dh.h>
+    #include <map>
+
+    // internal session context
+    // - ACTIVE means tl_session_setup succeeded
+    // - ACTIVE_VALIDATED means tl_session_setup succeeded, enclave_id is confirmed
+    //   _and_ we have verified chaincode_id & ecc_identity via ERCC
+    typedef tl_tlcc_state enum{INIT, ACTIVE, ACTIVE_VALIDATED, CLOSED} tl_tlcc_state_t;
+
+    struct tl_tlcc_ctx {
+        uint64_t session_id,
+        tl_tlcc_state_t state,
+        sgx_dh_session_t dh,
+        sgx_key_128bit_t session_key
+        char* channel_id,
+        channel_hash_t channel_hash,
+        char* cc,
+        char* enclave_id,
+        sgx_dh_session_enclave_identity_t ecc_identity,
+    } tl_tlcc_ctx_t;
+
+
+    // global variables
+    // - configured request processor
+    static tl_request_handler_t tl_req_handler;
+
+    static uint64_t next_session_id = 0;
+
+    // - in-flight sessions: map session-id -> context
+    static std::map<uint64_t, tl_tlcc_ctx_t> tl_sessions;
```

Comment:
TLCC_Enclave clearly needs some state so it can demultiplex between requests from different ECC_Enclaves and securely knows the corresponding context. By design this is also _ephemeral_ state as we want a session to last only between two instances of enclaves.  So the clean-up at tlcc is automatic if the peer and tlcc goes down.   Assuming that TLCC goes only down when the peer goes down (i.e., we ignore crash-recovery for now) and ECCs are also cleaned up when the peer goes down (as more or less what happens now?) then cleanup on the ECC side should also be fine.    

If we bring also crash-tolerance into the mix, then tlcc might at some point have to do garbage collection (e.g., to prevent running out of memory) and ecc would have to try to reconnect, including tx re-execution (or just abort current tx and expect submitter to re-submit) but that seems probably out-of-scope for now, i also don't think it would change the state or messages themselves, just the error handling logic (which is currently deferred to be defined when implementing it).

Reply:
> TLCC_Enclave clearly needs some state

It should be possible to make TLCC stateless by moving the session state to ECC. TLCC would just maintain a master secret, and derive per-ECC_enclave keys. It's non-trivial, but seems doable.

For the rest, we are on the same page. The error handling logic can be defined incrementally later.

Reply:
> > TLCC_Enclave clearly needs some state
> 
> It should be possible to make TLCC stateless by moving the session state to ECC. TLCC would just maintain a master secret, and derive per-ECC_enclave keys. It's non-trivial, but seems doable.

You are correct that one could attempt to outsource (encrypted and integrity-protected) state, similar to what some web-servers do to outsource session state to browsers via cookies.  Not 100% sure how you could guarantee the current session properties requiring only continuity of a peer enclave (which is very useful to guarantee view consistency).  But even if you could, i don't immediately see the benefit of doing so (and certainly adding some complexity: It's not that we will have o 10'000s of clients that storage would be a RAM issue or you have multiple servers, potentially with dynamic scale-out which would require state sharing, i.e., typical reasons why webservers do so?  Until we add crypto-paging, the tlcc enclave anyway has to be stateful?